modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -1054,6 +1067,206 @@ def _union_all(
+                    or (field.type.bit_width > cur_field.type.bit_width)
+                )
+            ):
+                col_fields[key] = field
+
+        if join == "outer":
+            for frame in frames:
+                table = frame.table
+                idx_width = 0 if frame.index_cols is None else len(frame.index_cols)
+                table_cols = table.column_names[idx_width:]
+                for col_name, table_col_name in zip(frame.columns, table_cols):
+                    add_col_field(table, col_name, table_col_name)
+        else:
+            col_names = {c for c in frames[0].columns} if len(frames) > 1 else []
+            if len(col_names) != 0:
+                for frame in frames[1:]:
+                    for name in list(col_names):
+                        if name not in frame.columns:
+                            col_names.remove(name)
+                if len(col_names) != 0:
+                    for frame in frames:
+                        table = frame.table
+                        idx_width = (
+                            0 if frame.index_cols is None else len(frame.index_cols)
+                        )
+                        table_cols = table.column_names[idx_width:]
+                        for col_name, table_col_name in zip(frame.columns, table_cols):
+                            if col_name in col_names:
+                                add_col_field(table, col_name, table_col_name)
+
+        if len(col_fields) == 0:
+            if ignore_index:
+                idx = RangeIndex(0, sum(len(f.table) for f in frames))
+            else:
+                idx = frames[0].index.append([f.index for f in frames[1:]])
+            idx_cols = mangle_index_names(idx.names)
+            idx_df = pd.DataFrame(index=idx).reset_index()
+            union = pyarrow.Table.from_pandas(idx_df).rename_columns(idx_cols)
+        else:
+            # Process empty frames with non-empty index
+            for frame in frames:
+                if len(frame.table) == 0 and len(frame.index) != 0:
+                    if ignore_index:
+                        frame.index = pd.RangeIndex(0, len(frame.index))
+                    idx = frame.index
+                    frame.index_cols = mangle_index_names(idx.names)
+                    idx_df = pd.DataFrame(index=idx).reset_index()
+                    frame.table = pyarrow.Table.from_pandas(idx_df)
+                    frame.table = frame.table.rename_columns(frame.index_cols)
+
+            idx_cols = None
+            idx_table = None
+            idx_ranges = None
+            idx_fields: typing.OrderedDict[str, pyarrow.Field] = OrderedDict()
+
+            if not ignore_index:
+                if all(f.index_cols is None for f in frames):
+                    # All frames have a trivial index
+                    idx_cols = mangle_index_names(self.index.names)
+                    idx_ranges = list(range(len(frames[0].table)))
```

Comment:
When you create `FrameData` objects, you materialize all indexes, even default ones. There is no need to create new ranges here then. Actually, it looks like you don't need any special handling of the case when all frames use default indexes. The `else` code below should handle it correctly by appending all indexes.

Reply:
I added the ranges in my initial implementation for performance optimization and it worked much faster. Now I see, it does not make any sense. Removed this code. Thanks.